core/src/main/scala/org/labrad/data/Data.scala

```diff
@@ -1208,12 +1232,72 @@ object Arr {
   def apply(elems: Seq[Data]): Data = apply(elems.toArray)
   def apply(elem: Data, elems: Data*): Data = apply(elem +: elems)
 
-  def apply(a: Array[Boolean]): Data = make[Boolean](a, TBool)
-  def apply(a: Array[Int]): Data = make[Int](a, TInt)
-  def apply(a: Array[Long]): Data = make[Long](a, TUInt)
+  def apply(a: Array[Boolean]): Data = {
+    val elem = TBool
+    val data = Data(TArr(elem, 1))
+    val m = a.length
+    data.setArrayShape(m)
+    val bytes = data.arrayBytes
+    var i = 0
+    while (i < m) {
+      bytes.setBool(elem.dataWidth * i, a(i))
+      i += 1
+    }
+    data
+  }
+  def apply(a: Array[Int]): Data = {
+    val elem = TInt
+    val data = Data(TArr(elem, 1))
+    val m = a.length
+    data.setArrayShape(m)
+    val bytes = data.arrayBytes
+    var i = 0
+    while (i < m) {
+      bytes.setInt(elem.dataWidth * i, a(i))
+      i += 1
+    }
+    data
+  }
+  def apply(a: Array[Long]): Data = {
+    val elem = TUInt
+    val data = Data(TArr(elem, 1))
+    val m = a.length
+    data.setArrayShape(m)
+    val bytes = data.arrayBytes
+    var i = 0
+    while (i < m) {
+      bytes.setUInt(elem.dataWidth * i, a(i))
+      i += 1
+    }
+    data
+  }
```

Comment:
I'd be really surprised if that works, because generally you can't match against a type like `Array[Int]` at runtime and I'd expect the compiler to warn about this. Usually this would require constraining the type parameter to tell the compiler to pass along extra information about the type, something like:
```scala
def arrayApply[T: ClassTag](a: Array[T], elem: Type): Data = { ... }
```
I also don't really like having a type check on every loop iteration, though I guess one could move the type check outside the loop, something like:
```scala
a match {
  case a: Array[Double] => while (i < m) { bytes.setDouble(elem.dataWidth * i, a(i)); i += 1 }
  ...
}
```
Anyway, we can try running some benchmarks to see what kind of performance hit we get.

Ideally we could use macros to avoid duplicating the code while still keeping things fast, but macros have their own complications, in particular the need for separate compilation which means the code that defines the macro has to live in a separate sbt project from any code that uses the macro (this was the reason for the core/manager split, but this gets to be a pain if we have to split things up even more finely). Alas, most non-macro generic mechanisms in scala come with performance hits...